be/src/olap/memtable.cpp

```diff
@@ -61,27 +60,55 @@ int MemTable::RowCursorComparator::operator()(const char* left, const char* righ
     return compare_row(lhs_row, rhs_row);
 }
 
-size_t MemTable::memory_usage() {
-    return _mem_pool->mem_tracker()->consumption();
-}
+void MemTable::insert(const Tuple* tuple) {
+	bool overwritten = false;
+    if (_keys_type == KeysType::DUP_KEYS) {
+        // Will insert directly, so use memory from _table_mem_pool
+        _tuple_buf = _table_mem_pool->allocate(_schema_size);
+        ContiguousRow row(_schema, _tuple_buf);
+        _tuple_to_row(tuple, &row, _table_mem_pool.get());
+        _skip_list->Insert((TableKey)_tuple_buf, &overwritten);
+        DCHECK(!overwritten) << "Duplicate key model meet overwrite in SkipList";
+        return;
+    }
 
-void MemTable::insert(Tuple* tuple) {
-    ContiguousRow row(_schema, _tuple_buf);
+    // For non-DUP models, for the data rows passed from the upper layer, when copying the data,
+    // we first allocate from _buffer_mem_pool, and then check whether it already exists in
+    // _skiplist.  If it exists, we aggregate the new row into the row in skiplist.
+    // otherwise, we need to copy it into _table_mem_pool before we can insert it.
+    _tuple_buf = _buffer_mem_pool->allocate(_schema_size);
+    ContiguousRow src_row(_schema, _tuple_buf);
+    _tuple_to_row(tuple, &src_row, _buffer_mem_pool.get());
+
+    bool is_exist = _skip_list->Find((TableKey)_tuple_buf, &_hint);
+    if (is_exist) {
+        _aggregate_two_row(src_row, _hint.curr->key);
+    } else {
+        _tuple_buf = _table_mem_pool->allocate(_schema_size);
+        ContiguousRow dst_row(_schema, _tuple_buf);
+        copy_row(&dst_row, src_row, _table_mem_pool.get());
+        _skip_list->InsertUseHint((TableKey)_tuple_buf, is_exist, &_hint);
+    }
 
+    // Make MemPool to be reusable, but does not free its memory
+    _buffer_mem_pool->clear();
+}
+
+void MemTable::_tuple_to_row(const Tuple* tuple, ContiguousRow* row, MemPool* mem_pool) {
     for (size_t i = 0; i < _slot_descs->size(); ++i) {
-        auto cell = row.cell(i);
+        auto cell = row->cell(i);
         const SlotDescriptor* slot = (*_slot_descs)[i];
 
         bool is_null = tuple->is_null(slot->null_indicator_offset());
-        void* value = tuple->get_slot(slot->tuple_offset());
-        _schema->column(i)->consume(&cell, (const char *)value, is_null, _mem_pool.get(), &_agg_object_pool);
+        const void* value = tuple->get_slot(slot->tuple_offset());
+        _schema->column(i)->consume(
+                &cell, (const char*)value, is_null, mem_pool, &_agg_object_pool);
```

Comment:
may be you forget _agg_object_pool

Reply:
object_pool is only used in `AggregateInfo`, so currently use only one `ObjectPool` object as a member of `Memtable`